### PR TITLE
Background async iterator for parallel processing of streams and chunking. 

### PR DESCRIPTION
**Pull request by hoytak**
This PR splits reading, chunking, and dedup into three separate threads, causing a 2-3X speedup for the clean filter.  Benchmarks now show 500-600 MB/s on larger files on an SSD. 

This PR implements several utilities to make this process testable and simple.

trait AsyncIterator<ErrorType>:  An async iterator type that is templated over the error type and Item to replace the very specific AsyncIterator in merkledb.  This is now defined in parutils.  The async_chunk_iterator was then modified to take and conform to this iterator trait.   

AsyncDataIterator trait -- Simply a specialization of the AsyncIterator trait to use ErrorType = GitXetRepoError and Item = Vec<u8>.   This is now the type used throughout the data_processing_* classes.  In addition, the existing AsyncFileIterator now conforms to this trait. 

BufferedAsyncIterator: An async iterator implementation that is initialized using any AsyncIterator-conforming object and a buffer size and spawns a background tokio thread to keep the iterator buffer full.  At the end, the original iterator can be retrieved using the finish() method.   Additionally, this iterator can be initialized using a collection of starting objects, meaning that we don't need AsyncIteratorWithPutback any more.  

Now, with these utilities, the code to background both the data loading and the chunking is quite simple: 

```
        // load enough data from reader to determine whether this file is just pass-through or not.
        let starting_data = { ... } 
        
        // Test if the file is pass through...  If it is return. 
        ...
        
        // Now, keep loading the data on a background thread.
        let raw_data_iter =
            BufferedAsyncIterator::new_with_starting_data(starting_data, reader, Some(512)).await?;

        // Pass this into the chunker to chunk on another background thread.
        let mut generator =
            BufferedAsyncIterator::new(async_chunk_target_default(raw_data_iter), Some(4096))
                .await?;
 ```

### CODE REVIEW

File: rust/merkledb/src/async_chunk_iterator.rs [hunk 1]
```diff
@@ -402,6 +399,24 @@ lazy_static! {
     };
 }
 
+// Annoying that we have to explicitly declare this, but that is the cost of using async_trait
```
Comment (hoytak) [inline rust/merkledb/src/async_chunk_iterator.rs:402]: Change to the reason being the use of Pin with a struct containing unsafe pointers.

File: rust/parutils/src/buffered_async_iterator.rs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+    /// is completed.  At that point, all items have been read into the local buffer.
+    pub fn new_detailed(
+        initial_data: Vec<It::Item>,
+        src_iter: It,
+        buffer_max_size: Option<usize>,
+    ) -> Self {
+        Self::new_impl(initial_data, src_iter, buffer_max_size, None)
+    }
+
+    /// Like new_deteailed, but also creates a tokio oneshot channel to return the src_iter object as
```
Comment (hoytak) [inline rust/parutils/src/buffered_async_iterator.rs:75]: typo

File: rust/merkledb/src/chunk_iterator.rs [hunk 3]
```diff
@@ -7,6 +7,25 @@ use std::cmp::min;
 use std::io;
 use std::io::Read;
 
+/*
+// For silly reasons, we have to implement this one here.
+impl<<E: Send + Sync + 'static, A> FallibleAsyncIterator<E> for A
```
Comment (hoytak) [inline rust/merkledb/src/chunk_iterator.rs:12]: Can be dropped -- not used anywhere.

File: rust/parutils/src/buffered_async_iterator.rs [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+use tracing::info;
+
+enum BufferItem<T: Send + Sync + 'static, E: Send + Sync + 'static> {
+    Value(T),
+    Completed,
+    Error(E),
+}
+
+/// Now, create a buffered stream.  The 'static lifetime specifiers essentially require the objects to be owned
+/// objects that don't contain any non-reference components.
```
Comment (seanses) [inline rust/parutils/src/buffered_async_iterator.rs:18]: "don't contain any non-reference" or "don't contain any reference"?

File: rust/parutils/src/buffered_async_iterator.rs [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+            };
+        }
+        Ok(ret)
+    }
+
+    /// If known, returns the number of items remaining.
+    fn items_remaining(&self) -> Option<usize> {
+        if self.all_items_in_buffer() {
+            let n = self.data_queue.len();
+            Some(n.saturating_sub(1))
```
Comment (seanses) [inline rust/parutils/src/buffered_async_iterator.rs:249]: A bit confused why subtracting 1 here?

File: rust/gitxetcore/src/lib.rs [hunk 6]
```diff
@@ -2,8 +2,7 @@
 
 extern crate more_asserts;
 
-mod async_file_iterator;
-mod async_iterator_with_putback;
+// mod async_iterator_with_putback;
```
Comment (seanses) [inline rust/gitxetcore/src/lib.rs:5]: Let's remove this

File: rust/parutils/src/buffered_async_iterator.rs (hunk 5 above)
Comment (seanses) [reply rust/parutils/src/buffered_async_iterator.rs:249]: Ah, it's because there's a `Complete` at the end. Let's put a comment here to explain.

**Review by seanses — APPROVED** [review]
LGTM! Let's fix the nits and get this in

**Pull request merged**